File: active_pilot/ios/Runner/DemoWebClient.swift
//
//  DemoWebClient.swift
//  coreapi-demo-simple
//
//  Created by Spencer Apsley on 2/6/18.
//  Copyright © 2018 Nokē Inc. All rights reserved.
//

import Foundation
import NokeMobileLibrary

let serverUrl = "https://coreapi-sandbox.appspot.com/"
let privateKey = "eyJhbGciOiJOT0tFX1BSSVZBVEVfU0FOREJPWCIsInR5cCI6IkpXVCJ9.eyJhbGciOiJOT0tFX1BSSVZBVEVfU0FOREJPWCIsImNvbXBhbnlfdXVpZCI6ImIxMDQwZTFhLTVmYTUtNGY3NC04ZGM3LTU1ZTBkZTlmMWVmMCIsImlzcyI6Im5va2UuY29tIn0.a076a54e54acf68a57ca4ff8446fafac73c9de72"

public protocol DemoWebClientDelegate{
    func didReceiveUnlockResponse(data: Data)
}

public class DemoWebClient: NSObject{
    
    public var delegate: DemoWebClientDelegate?
    static var sharedDemoWebClient: DemoWebClient?
    
    public static func shared()->DemoWebClient{
        if(sharedDemoWebClient == nil){
            sharedDemoWebClient = DemoWebClient.init()
        }
        return sharedDemoWebClient!
    }

    public func requestUnlock(noke: NokeDevice){
    
        let url = String.init(format:"%@%@", serverUrl, "unlock/")
        var jsonBody = [String: Any]()
        jsonBody["session"] = noke.session
        jsonBody["mac"] = noke.mac
        jsonBody["email"] = ""
    
        if(JSONSerialization.isValidJSONObject(jsonBody)){
            guard let jsonData = try? JSONSerialization.data(withJSONObject: jsonBody, options: JSONSerialization.WritingOptions.prettyPrinted) else{return}
            self.doRequest(url: url, jsonData: jsonData)
        }
    }
    
    public func requestUnshackle(noke: NokeDevice, email: String){
        
        let url = String.init(format:"%@%@", serverUrl, "unshackle/")
        var jsonBody = [String: Any]()
        jsonBody["session"] = noke.session
        jsonBody["mac"] = noke.mac
        jsonBody["email"] = email
        
        if(JSONSerialization.isValidJSONObject(jsonBody)){
            guard let jsonData = try? JSONSerialization.data(withJSONObject: jsonBody, options: JSONSerialization.WritingOptions.prettyPrinted) else{return}
            self.doRequest(url: url, jsonData: jsonData)
        }
    }
    
    internal func doRequest(url: String, jsonData: Data){
    
        var request = URLRequest(url: URL.init(string: url)!)
        request.httpMethod = "POST"
        request.httpBody = jsonData
        request.setValue("application/json", forHTTPHeaderField:"Content-Type")
        request.setValue("Bearer \(privateKey)", forHTTPHeaderField:"Authorization")
        let task = URLSession.shared.dataTask(with: request){data, response, error in
            guard let data = data, error == nil else{
                print("error=\(String(describing: error))")
                return
            }
        
            if let httpStatus = response as? HTTPURLResponse, httpStatus.statusCode != 200{
                print("statusCode should be 200, but is \(httpStatus.statusCode)")
                print("response = \(String(describing: response))")
            }
        
            let responseString = String(data: data, encoding: .utf8)
            print(responseString!)
            if(url.contains("unlock")){
                self.delegate?.didReceiveUnlockResponse(data: data)
            }
        }
        
        task.resume()
    }
    
    public func requestActivity(){
        
        let url = String.init(format:"%@%@", serverUrl, "activity/")
        let jsonBody = [String: Any]()
        
        if(JSONSerialization.isValidJSONObject(jsonBody)){
            guard let jsonData = try? JSONSerialization.data(withJSONObject: jsonBody, options: JSONSerialization.WritingOptions.prettyPrinted) else{return}
            self.doRequest(url: url, jsonData: jsonData)
        }
    }
}

File: active_pilot/android/app/src/main/kotlin/com/activepilot/aircraft/DemoWebClient.kt
package com.activepilot.aircraft

import android.content.Context
import android.content.pm.ApplicationInfo
import android.content.pm.PackageManager
import android.os.Bundle
import android.util.Log

import com.noke.nokemobilelibrary.NokeDevice

import org.json.JSONObject

import java.io.BufferedReader
import java.io.DataOutputStream
import java.io.IOException
import java.io.InputStream
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.URL
import java.security.KeyManagementException
import java.security.NoSuchAlgorithmException

import javax.net.ssl.SSLContext

/**
 * Created by Spencer on 2/7/18.
 * Demo Web Client for making requests to demo server and show API implementation
 *
 */

internal class DemoWebClient {
    private val serverUrl = "https://coreapi-beta.uc.r.appspot.com/"
    private var mDemoWebClientCallback: DemoWebClientCallback? = null

    fun requestUnlock(noke: NokeDevice, email: String) {
        /* Note: This is an example request to a demo server, it does not represent a request to the Noke Core API.
         * Requests should not be made to the Core API directly from the mobile app.
         * Please refer to the documentation for more details
         * (https://github.com/noke-inc/noke-mobile-library-android#nok%C4%93-mobile-library-for-android)
         */

        val thread = Thread(Runnable {
            try {

                val jsonObject = JSONObject()

                jsonObject.accumulate("session", noke.session)
                jsonObject.accumulate("mac", noke.mac)
                val url = serverUrl + "unlock/"

                Log.w(TAG, "JSON: $jsonObject")

                mDemoWebClientCallback?.onUnlockReceived(POST(url, jsonObject), noke)


            } catch (e: Exception) {
                e.printStackTrace()
            }
        })

        thread.start()
    }

    fun requestUnshackle(noke: NokeDevice, email: String) {
        /* Note: This is an example request to a demo server, it does not represent a request to the Noke Core API.
         * Requests should not be made to the Core API directly from the mobile app.
         * Please refer to the documentation for more details
         * (https://github.com/noke-inc/noke-mobile-library-android#nok%C4%93-mobile-library-for-android)
         */

        val thread = Thread(Runnable {
            try {

                val jsonObject = JSONObject()

                jsonObject.accumulate("session", noke.session)
                jsonObject.accumulate("mac", noke.mac)
                val url = serverUrl + "unshackle/"

                mDemoWebClientCallback?.onUnlockReceived(POST(url, jsonObject), noke)


            } catch (e: Exception) {
                e.printStackTrace()
            }
        })

        thread.start()
    }

    fun requestFobSync(noke: NokeDevice) {
        /* Note: This is an example request to a demo server, it does not represent a request to the Noke Core API.
         * Requests should not be made to the Core API directly from the mobile app.
         * Please refer to the documentation for more details
         * (https://github.com/noke-inc/noke-mobile-library-android#nok%C4%93-mobile-library-for-android)
         */

        val thread = Thread(Runnable {
            try {

                val jsonObject = JSONObject()

                jsonObject.accumulate("session", noke.session)
                jsonObject.accumulate("mac", noke.mac)
                val url = serverUrl + "fobs/sync/"

                mDemoWebClientCallback?.onUnlockReceived(POST(url, jsonObject), noke)


            } catch (e: Exception) {
                e.printStackTrace()
            }
        })

        thread.start()
    }


    fun setWebClientCallback(callback: DemoWebClientCallback) {
        this.mDemoWebClientCallback = callback
    }

    interface DemoWebClientCallback {

        fun onUnlockReceived(response: String, noke: NokeDevice)
    }

    companion object {

        private val TAG = DemoWebClient::class.java.simpleName
        private val privateKey = "eyJhbGciOiJOT0tFX1BSSVZBVEUiLCJ0eXAiOiJKV1QifQ.eyJhbGciOiJOT0tFX1BSSVZBVEUiLCJjb21wYW55X3V1aWQiOiIyZmQ2MGM1NS0xNGRmLTQxZWUtYWUyYS0wYjliOTY0OGQxNWUiLCJpc3MiOiJub2tlLmNvbSJ9.a1138da380ba229cf31d1ddd547e301e282492d8"

        private fun POST(urlStr: String, jsonObject: JSONObject): String {
            val conn: HttpURLConnection
            val inputStream: InputStream?
            var result = ""
            System.setProperty("http.keepAlive", "false")
            Log.w("CLIENT", "JSON: $jsonObject")

            try {
                val url = URL(urlStr)
                conn = url.openConnection() as HttpURLConnection

                //Create the SSL connection
                val sc: SSLContext
                sc = SSLContext.getInstance("TLS")
                sc.init(null, null, java.security.SecureRandom())

                //set Timeout and method
                conn.readTimeout = 60000
                conn.connectTimeout = 60000
                conn.requestMethod = "POST"
                conn.doInput = true
                conn.useCaches = false
                conn.setRequestProperty("Content-Type", "application/json")
                conn.setRequestProperty("Connection", "close")
                conn.setRequestProperty("charset", "utf-8")
                conn.setRequestProperty("Authorization", "Bearer $privateKey")

                val wr = DataOutputStream(conn.outputStream)
                wr.writeBytes(jsonObject.toString())
                wr.flush()
                wr.close()

                // Add any data you wish to post here
                conn.connect()
                inputStream = conn.inputStream

                if (inputStream != null) {
                    result = convertInputStreamToString(inputStream)
                } else {
                    result = "Did not work!"
                }
            } catch (e: IOException) {
                e.printStackTrace()
            } catch (e: NoSuchAlgorithmException) {
                e.printStackTrace()
            } catch (e: KeyManagementException) {
                e.printStackTrace()
            }

            return result
        }

        @Throws(IOException::class)
        private fun convertInputStreamToString(inputStream: InputStream): String {
            val bufferedReader = BufferedReader(InputStreamReader(inputStream))
            val result = StringBuilder()
            var line: String? = bufferedReader.readLine()
            while (line != null) {
                result.append(line)
                line = bufferedReader.readLine()
            }
            inputStream.close()
            return result.toString()
        }
    }
}